src/histogram.rs

```diff
@@ -436,6 +443,115 @@ pub fn duration_to_seconds(d: Duration) -> f64 {
     d.as_secs() as f64 + nanos
 }
 
+pub struct LocalHistogramCore {
+    histogram: Histogram,
+    counts: Vec<u64>,
+    count: u64,
+    sum: f64,
+}
+
+/// `LocalHistogram` is used for performance and in single-thread.
+/// Sometimes, if you very care the performance, you can use the LocalHistogram
+// and then flush the metric interval.
+#[derive(Clone)]
+pub struct LocalHistogram {
+    core: Rc<RefCell<LocalHistogramCore>>,
+}
+
+pub struct LocalHistogramTimer {
+    local: LocalHistogram,
+    start: Instant,
+}
+
+/// `LocalHistogramTimer` represents an event being timed. When the timer goes out of
+/// scope, the duration will be observed, or call `observe_duration` to manually
+/// observe.
+///
+/// NOTICE: A timer can be observed only once (automatically or manually).
+impl LocalHistogramTimer {
+    /// `observe_duration` observes the amount of time in seconds since
+    /// `LocalHistogram.start_timer` was called.
+    pub fn observe_duration(self) {
+        drop(self);
+    }
+
+    fn observe(&mut self) {
+        let v = duration_to_seconds(self.start.elapsed());
+        self.local.observe(v)
+    }
+}
+
+impl<'a> Drop for LocalHistogramTimer {
+    fn drop(&mut self) {
+        self.observe()
+    }
+}
+
+impl LocalHistogramCore {
+    fn new(histogram: Histogram) -> LocalHistogramCore {
+        let counts = vec![0;histogram.core.counts.len()];
+
+        LocalHistogramCore {
+            histogram: histogram,
+            counts: counts,
+            count: 0,
+            sum: 0.0,
+        }
+    }
+
+    pub fn observe(&mut self, v: f64) {
+        // Try find the bucket.
+        let mut iter =
+            self.histogram.core.upper_bounds.iter().enumerate().filter(|&(_, f)| v <= *f);
+        if let Some((i, _)) = iter.next() {
+            self.counts[i] += 1;
+        }
+
+        self.count += 1;
+        self.sum += v;
+    }
+
+    pub fn flush(&mut self) {
```

Comment:
API suggestion: the usual case will be to `flush`, so make that easy. Instead of requiring user to call `flush`, make it happen automatically via `Drop`. Add a method to forget the local changes.